### ISSUE TITLE
Support Nats java agent instrumentation

### ISSUE THREAD

**Issue by pkgonan** [open-telemetry/opentelemetry-java-instrumentation#9397]
### Is your feature request related to a problem? Please describe.

Nats is a very fast and scalable Pub/Sub messaging system that replaces WebSockets.

We are using Otel APM, but the metrics that communicate with Nats are not being exposed.

Therefore, we are requesting support for OTEL java agent instrumentation for Nats Client.

Reference
* https://github.com/nats-io/nats.java

### Describe the solution you'd like

Development new java agent instrumentation for java nats client.

**Comment by laurit** [comment]
Hey @pkgonan,

We likely won't have time and resources to implement this instrumentation ourselves, but we would find an external contribution welcome.

**Comment by wl2027** [comment]
@laurit It seems that this isuess has been open for a long time, it can be assigned to me, I will try my best to complete it.

**Comment by laurit** [comment]
> @laurit It seems that this isuess has been open for a long time, it can be assigned to me, I will try my best to complete it.

sure, have fun

**Comment by AlixBa** [comment]
@laurit I am accumulating commits [here](https://github.com/open-telemetry/opentelemetry-java-instrumentation/compare/main...AlixBa:opentelemetry-java-instrumentation:instrument-dispatcher?expand=1), and I think I almost reach a point where it feels like the bare minimum is supported (publish, subscribe, jet stream excluded). 

What would be the best for a review? All-in or split by block? Is there anything to do first as this would be a new module?

**Comment by laurit** [comment]
> What would be the best for a review? All-in or split by block? Is there anything to do first as this would be a new module?

For review it is best to have a fully functional instrumentation so we could see what telemetry it creates and compare it with existing messaging instrumentations.

**Comment by weilhuang** [comment]
> hey [@wl2027](https://github.com/wl2027) did you start anything on this topic? I'd like to work on the library instrumentation (not the agent part) if there is a basic structure available.

I was too busy with work recently and missed this message. Since I had already completed part of the work before, I feel confused about the support for instrumentation starting from version 2.21.0, because NATS had already supported message headers passing in version [2.10.0](https://github.com/nats-io/nats.java/releases/tag/2.10.0)

**Comment by AlixBa** [comment]
Then I guess we're not talking about the same thing. Unless I completely missed it, I don't see any instrumentation for [this](https://opentelemetry.io/docs/specs/semconv/messaging/messaging-spans/). I was referring to creating spans & traces for publication and subscriptions over NATS (with or without JetStreams).

**Comment by weilhuang** [comment]
> Then I guess we're not talking about the same thing. Unless I completely missed it, I don't see any instrumentation for [this](https://opentelemetry.io/docs/specs/semconv/messaging/messaging-spans/). I was referring to creating spans & traces for publication and subscriptions over NATS (with or without JetStreams).

You're right. Implementing instrumentation for NATS publish/subscribe operations is indeed the primary objective here. I'll be happy to contribute additional observability support for other NATS features once this core instrumentation is completed. Thank you for your willingness to participate and contribute to this effort.

**Comment by shubhajyoti-bagchi** [comment]
Hello @AlixBa @wl2027, is the work for NATS core instrumentation still active ?

**Comment by AlixBa** [comment]
Hey @shubhajyoti-bagchi 

Sorry I haven't been able to rebase on the latest changes (since July) to submit it to approval. The MR is almost ready as soon as I'll get some spare time to include latest changes. [Other than this, it's ready, yes](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/13999)

**Comment by shubhajyoti-bagchi** [comment]
Sure, thanks for the reply @AlixBa
